stan/math/rev/core/grad.hpp

```diff
@@ -37,6 +37,26 @@ static void grad(Vari* vi) {
   }
 }
 
+/**
+ * Compute the gradient for all variables starting from the end of the AD tape.
+ * This function does not recover memory.  The chain
+ * rule is applied working down the stack from the last vari created on the
+ * AD tape and then calling each vari's `chain()` method in turn.
+ *
+ * <p>This function computes a nested gradient only going back as far
+ * as the last nesting.
+ *
+ * <p>This function does not recover any memory from the computation.
+ *
+ */
+static void grad() {
```

Comment:
The duplicate code here can be eliminated pretty easily.

Either call this grad inside the other grad or do a default nullptr argument in the other thing:

```cpp
template <typename Vari>
static void grad(Vari* vi) {
  vi->init_dependent();
  grad();
}
```

or

```cpp
template <typename Vari>
static void grad(Vari* vi = nullptr) {
  if(!vi)
    vi->init_dependent();
  size_t end = ChainableStack::instance_->var_stack_.size();
  size_t beginning = empty_nested() ? 0 : end - nested_size();
  for (size_t i = end; i-- > beginning;) {
    ChainableStack::instance_->var_stack_[i]->chain();
  }
}
```

and delete the second definition.

Reply:
Done deal!